libs/langchain/langchain/document_loaders/blob_loaders/schema.py

```diff
@@ -28,22 +28,36 @@ class Blob(BaseModel):
     Inspired by: https://developer.mozilla.org/en-US/docs/Web/API/Blob
     """
 
-    data: Union[bytes, str, None]  # Raw data
-    mimetype: Optional[str] = None  # Not to be confused with a file extension
-    encoding: str = "utf-8"  # Use utf-8 as default encoding, if decoding to string
-    # Location where the original content was found
-    # Represent location on the local file system
-    # Useful for situations where downstream code assumes it must work with file paths
-    # rather than in-memory content.
+    data: Union[bytes, str, None]
+    """Raw data associated with the blob."""
+    mimetype: Optional[str] = None
+    """MimeType not to be confused with a file extension."""
+    encoding: str = "utf-8"
+    """Encoding to use if decoding the bytes into a string.
+    
+    Use utf-8 as default encoding, if decoding to string.
+    """
     path: Optional[PathLike] = None
+    """Location where the original content was found."""
+
+    metadata: Dict[str, Any] = Field(default_factory=dict)
+    """Metadata about the blob (e.g., source)"""
 
     class Config:
         arbitrary_types_allowed = True
         frozen = True
 
     @property
     def source(self) -> Optional[str]:
-        """The source location of the blob as string if known otherwise none."""
+        """The source location of the blob as string if known otherwise none.
+
+        If a path is associated with the blob, it will default to the path location.
+
+        Unless explicitly set via a metadata field called "source", in which
+        case that value will be used instead.
+        """
+        if self.metadata and "source" in self.metadata:
+            return cast(Optional[str], self.metadata["source"])
```

Comment:
could this break stuff?

Reply:
Potentially, but more likely to fix things rather than break them -- folks usually want the ultimate provenance (original url) rather than the immediate one (e.g., temp file) and have been modifying loaders/parsers to get this logic in.